File: typescript/src/functors/sql.ts
/**
 * SQL extension functors for JSE.
 *
 * Migrated from the original engine implementation:
 * - $pattern: Generate SQL for triple pattern matching
 * - $query: Generate SQL for multi-pattern queries
 */

import type { JseValue, AstNodeLike } from "../types.js";
import { Env } from "../env.js";
import { Parser } from "../ast/parser.js";

/**
 * Query field list for SQL SELECT.
 */
export const QUERY_FIELDS = "subject, predicate, object, meta";

/**
 * Convert $pattern arguments to PostgreSQL jsonb containment triple.
 * - ["$pattern", "$*", "author of", "$*"] -> triple: ["author of"] (filters $*)
 * - ["$pattern", "Liu Xin", "author of", "$*"] -> triple: ["Liu Xin", "author of"]
 */
export function patternToTriple(
  subject: string,
  predicate: string,
  object: string
): unknown[] {
  const pattern: unknown[] = [];
  // Filter out $* wildcard
  if (subject !== "$*" && subject !== "") {
    pattern.push(subject);
  }
  if (predicate !== "$*" && predicate !== "") {
    pattern.push(predicate);
  }
  if (object !== "$*" && object !== "") {
    pattern.push(object);
  }
  return pattern;
}

/**
 * Convert $pattern arguments for $query's local environment.
 * - ["pattern", "$*", "author of", "$*"] -> triple: ["*", "author of", "*"] (expands $*)
 * - ["pattern", "Liu Xin", "author of", "$*"] -> triple: ["Liu Xin", "author of", "*"]
 */
function patternToTripleForQuery(
  subject: string,
  predicate: string,
  object: string
): unknown[] {
  const pattern: unknown[] = [];
  // Expand $* wildcard to "*" string
  if (subject === "$*") {
    pattern.push("*");
  } else if (subject !== "") {
    pattern.push(subject);
  }
  if (predicate === "$*") {
    pattern.push("*");
  } else if (predicate !== "") {
    pattern.push(predicate);
  }
  if (object === "$*") {
    pattern.push("*");
  } else if (object !== "") {
    pattern.push(object);
  }
  return pattern;
}

/**
 * Build SQL WHERE clause for a triple pattern.
 */
export function tripleToSqlCondition(triple: unknown[]): string {
  const json = JSON.stringify({ triple });
  // Add spaces after colons for readability (matching Python output)
  const spaced = json.replace(/:/g, ": ");
  const escaped = spaced.replace(/'/g, "''");
  return `meta @> '${escaped}'`;
}

// Type alias for functors
export type Functor = (env: Env, ...args: JseValue[]) => JseValue;

/**
 * Generate SQL for triple pattern matching.
 * Form: [$pattern, subject, predicate, object]
 */
export function _pattern(env: Env, ...args: JseValue[]): JseValue {
  if (args.length < 3) {
    throw new Error("$pattern requires (subject, predicate, object)");
  }

  const subj = env.eval(args[0]);
  const pred = env.eval(args[1]);
  const obj = env.eval(args[2]);

  if (
    typeof subj !== "string" ||
    typeof pred !== "string" ||
    typeof obj !== "string"
  ) {
    throw new Error("$pattern requires string arguments");
  }

  const triple = patternToTriple(subj, pred, obj);
  const cond = tripleToSqlCondition(triple);

  return cond
}

/**
 * Expression evaluation pass-through.
 * Form: [$expr, expression] or {$expr: expression}
 * Evaluates the expression and returns the result.
 */
export function _expr(env: Env, ...args: JseValue[]): JseValue {
  if (args.length === 0) {
    return null;
  }
  return env.eval(args[0]);
}

/**
 * SQL-specific AND: joins conditions with " and ".
 * This is LOCAL-ONLY for $query, different from logical _and in utils.ts.
 */
function _and(env: Env, ...args: JseValue[]): JseValue {
  const tokens = args.map((e) => env.eval(e));
  return tokens.join(" and ");
}

/**
 * Wildcard helper for local scope.
 */
function _wildcard(_env: Env, ..._args: JseValue[]): JseValue {
  return "*";
}

/**
 * Local $pattern for $query's environment.
 * Expands $* to "*" in most cases, but filters when all args are $*.
 */
function _patternForQuery(env: Env, ...args: JseValue[]): JseValue {
  if (args.length < 3) {
    throw new Error("$pattern requires (subject, predicate, object)");
  }

  const subj = env.eval(args[0]);
  const pred = env.eval(args[1]);
  const obj = env.eval(args[2]);

  if (
    typeof subj !== "string" ||
    typeof pred !== "string" ||
    typeof obj !== "string"
  ) {
    throw new Error("$pattern requires string arguments");
  }

  // Special case: if all arguments are $*, filter them out
  if (subj === "$*" && pred === "$*" && obj === "$*") {
    const triple = patternToTriple(subj, pred, obj);  // Filters $*
    const cond = tripleToSqlCondition(triple);
    return cond;
  }

  // Otherwise, expand $* to "*"
  const triple = patternToTripleForQuery(subj, pred, obj);
  const cond = tripleToSqlCondition(triple);
  return cond;
}

/**
 * Generate SQL for multi-pattern query.
 * Form: [$query, condition]
 * where condition is an AST expression with local operators ($pattern, $and, $*)
 */
export function _query(env: Env, ...args: JseValue[]): JseValue {
  // Create local environment inheriting from parent
  const local = new Env(env);
  local.load({
    "$pattern": _patternForQuery,  // Use local version that expands $*
    "$and": _and, // SQL-specific AND
    "$*": _wildcard,
  });

  const parser = new Parser(local);
  // Parse all args as a single expression (like Python's tuple unpacking)
  const condition = parser.parse(args) as AstNodeLike;
  const where = condition.apply(local); // Evaluate in local env

  return `select ${QUERY_FIELDS} \nfrom statement \nwhere \n    ${where} \noffset 0\nlimit 100 \n`;
}

// Dict of all SQL functors for registration
export const SQL_FUNCTORS: Record<string, Functor> = {
  $query: _query,
};

File: rust/src/functors/sql.rs
//! SQL extension functors for JSE.
//!
//! Following Python's functors/sql.py pattern:

use std::rc::Rc;
use std::cell::RefCell;
use serde_json::Value;
use crate::env::{Env, Functor};
use crate::ast::AstError;
use crate::sql::{pattern_to_triple, triple_to_sql_condition};

pub const QUERY_FIELDS: &str = "subject, predicate, object, meta";

pub fn pattern(env: &Rc<RefCell<Env>>, args: &[Value]) -> Result<Value, AstError> {
    if args.len() < 3 {
        return Err(AstError::ArityError("$pattern requires (subject, predicate, object)".to_string()));
    }

    let subj = args[0].as_str()
        .ok_or_else(|| AstError::TypeError("$pattern requires string arguments".to_string()))?;
    let pred = args[1].as_str()
        .ok_or_else(|| AstError::TypeError("$pattern requires string arguments".to_string()))?;
    let obj = args[2].as_str()
        .ok_or_else(|| AstError::TypeError("$pattern requires string arguments".to_string()))?;

    let triple = pattern_to_triple(subj, pred, obj);
    let cond = triple_to_sql_condition(&triple);

    let sql = format!(
        "select \n    subject, predicate, object, meta \nfrom statement as s \nwhere {} \noffset 0\nlimit 100 \n",
        cond
    );

    Ok(Value::String(sql))
}

pub fn expr(env: &Rc<RefCell<Env>>, args: &[Value]) -> Result<Value, AstError> {
    if args.is_empty() {
        return Ok(Value::Null);
    }
    Ok(args[0].clone())
}

pub fn query(env: &Rc<RefCell<Env>>, args: &[Value]) -> Result<Value, AstError> {
    if args.len() < 1 {
        return Err(AstError::ArityError("$query expects [op, patterns array] or [patterns array]".to_string()));
    }

    // Helper function to evaluate an expression
    fn eval_expr(env: &Rc<RefCell<Env>>, expr: &Value) -> Result<Value, AstError> {
        // Re-parse and evaluate the expression
        let parser = crate::ast::Parser::new(Rc::clone(env));
        let ast = parser.parse(expr)?;
        ast.apply(env)
    }

    // Extract patterns from raw JSON structure
    // Don't evaluate the top-level array yet
    let raw_value = &args[0];

    // Check if this is [op, patterns] format or just [patterns]
    let (patterns_array, has_op) = match raw_value {
        Value::Array(arr) if !arr.is_empty() => {
            // Check if first element is an operator
            if arr[0].is_string() {
                let s = arr[0].as_str().unwrap();
                if s.starts_with('$') && s != "$*" && !s.starts_with("$$") && s != "$expr" && s != "$query" && s != "$pattern" {
                    // This is [op, patterns] format
                    // Check if it's a logical operator that needs special handling
                    if s == "$and" || s == "$or" {
                        // For $and/$or with patterns, the patterns are in the second element
                        if let Some(patterns) = arr.get(1) {
                            (patterns, true)
                        } else {
                            (raw_value, false)
                        }
                    } else {
                        // Unknown operator - treat whole thing as patterns
                        (raw_value, false)
                    }
                } else {
                    // First element is not an operator - this is [patterns]
                    (raw_value, false)
                }
            } else {
                // First element is not a string - this is [patterns]
                (raw_value, false)
            }
        }
        _ => (raw_value, false),
    };

    let patterns = patterns_array.as_array()
        .ok_or_else(|| AstError::TypeError("$query argument must be an array".to_string()))?;

    let mut conditions = Vec::new();
    for pattern_val in patterns {
        // Evaluate each pattern expression
        let sql_val = eval_expr(env, pattern_val)?;
        let sql = sql_val.as_str()
            .ok_or_else(|| AstError::TypeError("Pattern must evaluate to SQL string".to_string()))?;

        // Extract WHERE clause
        let re = regex::RegexBuilder::new(r"where\s+(.+?)\s+offset")
            .case_insensitive(true)
            .dot_matches_new_line(true)
            .build()
            .map_err(|e| AstError::EvalError(e.to_string()))?;

        let cond = if let Some(cap) = re.captures(sql) {
            format!("({})", cap.get(1).unwrap().as_str().trim())
        } else {
            sql.to_string()
        };
        conditions.push(cond);
    }

    let where_clause = conditions.join(" and \n    ");
    let sql = format!(
        "select {} \nfrom statement \nwhere \n    {} \noffset 0\nlimit 100 \n",
        QUERY_FIELDS, where_clause
    );

    Ok(Value::String(sql))
}

pub fn sql_functors() -> std::collections::HashMap<&'static str, Functor> {
    let mut m = std::collections::HashMap::new();
    m.insert("$pattern", pattern as Functor);
    m.insert("$expr", expr as Functor);
    m.insert("$query", query as Functor);
    m
}
